<html><head></head><body>
<div class="IMG---Figure" id="_idContainer051">
<h1 class="chapter-number" id="_idParaDest-182"><a id="_idTextAnchor183"/><span class="koboSpan" id="kobo.1.1">7</span></h1>
<h1 id="_idParaDest-183"><a id="_idTextAnchor184"/><span class="koboSpan" id="kobo.2.1">Error Handling and Logging</span></h1>
<p><span class="koboSpan" id="kobo.3.1">We described how Python and Go run and how they access the network in the previous chapter; however, we missed two important points when building our network automation solution: how we report program execution events and how we </span><span class="No-Break"><span class="koboSpan" id="kobo.4.1">handle errors.</span></span></p>
<p><span class="koboSpan" id="kobo.5.1">These two topics are not as easy as they seem, and they are, most of the time, implemented in the system poorly. </span><span class="koboSpan" id="kobo.5.2">Some network developers might not do it properly because of lack of knowledge, but there are also some developers that don’t do it properly because of time constraints and extra time needed </span><span class="No-Break"><span class="koboSpan" id="kobo.6.1">for coding.</span></span></p>
<p><span class="koboSpan" id="kobo.7.1">But are these activities really important? </span><span class="koboSpan" id="kobo.7.2">Let’s examine these in this chapter. </span><span class="koboSpan" id="kobo.7.3">First, let’s investigate how and why we handle errors and then why and how we do </span><span class="No-Break"><span class="koboSpan" id="kobo.8.1">event logging.</span></span></p>
<p><span class="koboSpan" id="kobo.9.1">Here are the topics we are going to cover in </span><span class="No-Break"><span class="koboSpan" id="kobo.10.1">this chapter:</span></span></p>
<ul>
<li><span class="koboSpan" id="kobo.11.1">Writing code for </span><span class="No-Break"><span class="koboSpan" id="kobo.12.1">error handling</span></span></li>
<li><span class="No-Break"><span class="koboSpan" id="kobo.13.1">Logging events</span></span></li>
<li><span class="koboSpan" id="kobo.14.1">Adding logging to </span><span class="No-Break"><span class="koboSpan" id="kobo.15.1">your code</span></span></li>
</ul>
<p><span class="koboSpan" id="kobo.16.1">After reading this chapter, you will be able to add effectively code to handle errors and log events in your </span><span class="No-Break"><span class="koboSpan" id="kobo.17.1">network development.</span></span></p>
<h1 id="_idParaDest-184"><a id="_idTextAnchor185"/><span class="koboSpan" id="kobo.18.1">Technical requirements</span></h1>
<p><span class="koboSpan" id="kobo.19.1">The source code described in this chapter is stored in the GitHub repository </span><span class="No-Break"><span class="koboSpan" id="kobo.20.1">at </span></span><a href="https://github.com/PacktPublishing/Network-Programming-and-Automation-Essentials/tree/main/Chapter07"><span class="No-Break"><span class="koboSpan" id="kobo.21.1">https://github.com/PacktPublishing/Network-Programming-and-Automation-Essentials/tree/main/Chapter07</span></span></a><span class="No-Break"><span class="koboSpan" id="kobo.22.1">.</span></span></p>
<h1 id="_idParaDest-185"><a id="_idTextAnchor186"/><span class="koboSpan" id="kobo.23.1">Writing code for error handling</span></h1>
<p><span class="koboSpan" id="kobo.24.1">To see how important</span><a id="_idIndexMarker846"/><span class="koboSpan" id="kobo.25.1"> handling errors is, we have to think of our system as a whole, including inputs and output. </span><span class="koboSpan" id="kobo.25.2">Our code by itself might never experience an error; however, when integrated with other systems, it might cause unpredictable outputs, or might just crash and </span><span class="No-Break"><span class="koboSpan" id="kobo.26.1">stop working.</span></span></p>
<p><span class="koboSpan" id="kobo.27.1">Therefore, handling errors is important to cope with the unpredictability of inputs and protect your code to avoid wrong outputs or crashes. </span><span class="koboSpan" id="kobo.27.2">But how do we </span><span class="No-Break"><span class="koboSpan" id="kobo.28.1">do that?</span></span></p>
<p><span class="koboSpan" id="kobo.29.1">First, we need to identify</span><a id="_idIndexMarker847"/><span class="koboSpan" id="kobo.30.1"> our inputs, and then we create a series of different combinations of values that are sent to our inputs. </span><span class="koboSpan" id="kobo.30.2">The behavior of these input combinations are then evaluated by running our code. </span><span class="koboSpan" id="kobo.30.3">For a function, we do that by adding unit tests, as we discussed in </span><a href="B18165_05.xhtml#_idTextAnchor127"><span class="No-Break"><em class="italic"><span class="koboSpan" id="kobo.31.1">Chapter 5</span></em></span></a><span class="koboSpan" id="kobo.32.1">. </span><span class="koboSpan" id="kobo.32.2">For the system, we add integration tests and end-to-end testing. </span><span class="koboSpan" id="kobo.32.3">Additional techniques were also discussed in </span><a href="B18165_05.xhtml#_idTextAnchor127"><span class="No-Break"><em class="italic"><span class="koboSpan" id="kobo.33.1">Chapter 5</span></em></span></a><span class="No-Break"><span class="koboSpan" id="kobo.34.1">.</span></span></p>
<p><span class="koboSpan" id="kobo.35.1">But what is the correct way to write code for handling errors? </span><span class="koboSpan" id="kobo.35.2">It will depend on </span><span class="No-Break"><span class="koboSpan" id="kobo.36.1">the language.</span></span></p>
<p><span class="koboSpan" id="kobo.37.1">Writing code to handle errors is quite different in Go compared to Python. </span><span class="koboSpan" id="kobo.37.2">Let’s see how we do this effectively in Go then </span><span class="No-Break"><span class="koboSpan" id="kobo.38.1">in Python.</span></span></p>
<h2 id="_idParaDest-186"><a id="_idTextAnchor187"/><span class="koboSpan" id="kobo.39.1">Adding error handling in Go</span></h2>
<p><span class="koboSpan" id="kobo.40.1">The design of the Go language</span><a id="_idIndexMarker848"/><span class="koboSpan" id="kobo.41.1"> requires explicitly checking errors when they occur, which is different from throwing exceptions and then catching them like in Python. </span><span class="koboSpan" id="kobo.41.2">In Go, errors are just values returned by functions, which makes Go coding a bit more </span><a id="_idIndexMarker849"/><span class="koboSpan" id="kobo.42.1">verbose and perhaps more repetitive. </span><span class="koboSpan" id="kobo.42.2">With Python, you would not need to check errors, as an exception would be raised, but in Go, you have </span><a id="_idIndexMarker850"/><span class="koboSpan" id="kobo.43.1">to check the errors. </span><span class="koboSpan" id="kobo.43.2">But on the other hand, Go error handling is much simpler compared </span><span class="No-Break"><span class="koboSpan" id="kobo.44.1">to Python.</span></span></p>
<p><span class="koboSpan" id="kobo.45.1">Errors in Go are created by using an </span><strong class="source-inline"><span class="koboSpan" id="kobo.46.1">error</span></strong><span class="koboSpan" id="kobo.47.1"> type interface, </span><span class="No-Break"><span class="koboSpan" id="kobo.48.1">as follows:</span></span></p>
<pre class="source-code"><span class="koboSpan" id="kobo.49.1">
type error interface {
    Error() string
}</span></pre>
<p><span class="koboSpan" id="kobo.50.1">As you can see in the preceding code, the error implementation in Go is quite simple by using a method called </span><strong class="source-inline"><span class="koboSpan" id="kobo.51.1">Error()</span></strong><span class="koboSpan" id="kobo.52.1"> that returns an error message as </span><span class="No-Break"><span class="koboSpan" id="kobo.53.1">a string.</span></span></p>
<p><span class="koboSpan" id="kobo.54.1">The correct way to construct an error in your code is by using either the </span><strong class="source-inline"><span class="koboSpan" id="kobo.55.1">errors</span></strong><span class="koboSpan" id="kobo.56.1"> or </span><strong class="source-inline"><span class="koboSpan" id="kobo.57.1">fmt</span></strong> <span class="No-Break"><span class="koboSpan" id="kobo.58.1">standard libraries</span></span></p>
<p><span class="koboSpan" id="kobo.59.1">The following are two examples of using each of them for a </span><span class="No-Break"><span class="koboSpan" id="kobo.60.1">division-by-zero function.</span></span></p>
<p><span class="koboSpan" id="kobo.61.1">Using the </span><strong class="source-inline"><span class="koboSpan" id="kobo.62.1">errors</span></strong><span class="koboSpan" id="kobo.63.1"> library looks </span><span class="No-Break"><span class="koboSpan" id="kobo.64.1">as follows:</span></span></p>
<pre class="source-code"><span class="koboSpan" id="kobo.65.1">
func divide(q int, d int) (int, error) {
    if d == 0 {
        return 0, errors.New("division by zero not valid")
    }
    return q / d, nil
}</span></pre>
<p><span class="koboSpan" id="kobo.66.1">Using the </span><strong class="source-inline"><span class="koboSpan" id="kobo.67.1">fmt</span></strong><span class="koboSpan" id="kobo.68.1"> library</span><a id="_idIndexMarker851"/><span class="koboSpan" id="kobo.69.1"> looks </span><span class="No-Break"><span class="koboSpan" id="kobo.70.1">as</span></span><span class="No-Break"><a id="_idIndexMarker852"/></span><span class="No-Break"><span class="koboSpan" id="kobo.71.1"> follows:</span></span></p>
<pre class="source-code"><span class="koboSpan" id="kobo.72.1">
func divide(q int, d int) (int, error) {
    if d == 0 {
        return 0, fmt.Errorf("divided by zero not valid")
    }
    return q / d, nil
}</span></pre>
<p><span class="koboSpan" id="kobo.73.1">The preceding two examples produce the same results. </span><span class="koboSpan" id="kobo.73.2">For instance, if you call any of the two functions with </span><strong class="source-inline"><span class="koboSpan" id="kobo.74.1">fmt.Println(divide(10, 0))</span></strong><span class="koboSpan" id="kobo.75.1">, it will print the following output: </span><strong class="source-inline"><span class="koboSpan" id="kobo.76.1">0 divided by zero </span></strong><span class="No-Break"><strong class="source-inline"><span class="koboSpan" id="kobo.77.1">not valid</span></strong></span><span class="No-Break"><span class="koboSpan" id="kobo.78.1">.</span></span></p>
<p><span class="koboSpan" id="kobo.79.1">The major difference </span><a id="_idIndexMarker853"/><span class="koboSpan" id="kobo.80.1">between </span><strong class="source-inline"><span class="koboSpan" id="kobo.81.1">fmt.Errorf</span></strong><span class="koboSpan" id="kobo.82.1"> and </span><strong class="source-inline"><span class="koboSpan" id="kobo.83.1">errors.New</span></strong><span class="koboSpan" id="kobo.84.1"> is the possibility of formatting the string and adding values. </span><span class="koboSpan" id="kobo.84.2">Another point is that </span><strong class="source-inline"><span class="koboSpan" id="kobo.85.1">errors.New</span></strong><span class="koboSpan" id="kobo.86.1"> is faster because it does not invoke </span><span class="No-Break"><span class="koboSpan" id="kobo.87.1">the formatter.</span></span></p>
<p><span class="koboSpan" id="kobo.88.1">If you are looking to create custom errors, stack traces, and more advanced error features, consider using the </span><strong class="source-inline"><span class="koboSpan" id="kobo.89.1">errors</span></strong><span class="koboSpan" id="kobo.90.1"> library or third-party libraries such as the popular </span><strong class="source-inline"><span class="koboSpan" id="kobo.91.1">pkg/errors</span></strong><span class="koboSpan" id="kobo.92.1"> (</span><a href="https://pkg.go.dev/github.com/pkg/errors"><span class="koboSpan" id="kobo.93.1">https://pkg.go.dev/github.com/pkg/errors</span></a><span class="koboSpan" id="kobo.94.1">) or </span><span class="No-Break"><strong class="source-inline"><span class="koboSpan" id="kobo.95.1">golang.org/x/xerrors</span></strong></span><span class="No-Break"><span class="koboSpan" id="kobo.96.1"> (</span></span><a href="https://pkg.go.dev/golang.org/x/xerrors"><span class="No-Break"><span class="koboSpan" id="kobo.97.1">https://pkg.go.dev/golang.org/x/xerrors</span></span></a><span class="No-Break"><span class="koboSpan" id="kobo.98.1">).</span></span></p>
<p><span class="koboSpan" id="kobo.99.1">Let’s focus now on the best practices when writing code for error handling in Go. </span><span class="koboSpan" id="kobo.99.2">The following are the top </span><span class="No-Break"><span class="koboSpan" id="kobo.100.1">best practices.</span></span></p>
<h3><span class="koboSpan" id="kobo.101.1">Return errors last and values to 0</span></h3>
<p><span class="koboSpan" id="kobo.102.1">When creating a function that returns several values, the error should be placed as the last argument returned. </span><span class="koboSpan" id="kobo.102.2">When returning </span><a id="_idIndexMarker854"/><span class="koboSpan" id="kobo.103.1">values with errors, use 0 when it is a number and </span><strong class="source-inline"><span class="koboSpan" id="kobo.104.1">empty string</span></strong><span class="koboSpan" id="kobo.105.1"> for strings, as in the </span><span class="No-Break"><span class="koboSpan" id="kobo.106.1">following example:</span></span></p>
<pre class="source-code"><span class="koboSpan" id="kobo.107.1">
func findNameCount(text string) (string, int, error) {
    if len(text) &lt; 5 {
        return "", 0, fmt.Errorf("text too small")
    }
    . </span><span class="koboSpan" id="kobo.107.2">. </span><span class="koboSpan" id="kobo.107.3">.
</span><span class="koboSpan" id="kobo.107.4">}</span></pre>
<p><span class="koboSpan" id="kobo.108.1">In the preceding example, the value of the string returned is empty and the value of </span><strong class="source-inline"><span class="koboSpan" id="kobo.109.1">int</span></strong><span class="koboSpan" id="kobo.110.1"> returned is 0. </span><span class="koboSpan" id="kobo.110.2">These values are just suggestions, because when an error is returned, the calling statement will first check whether there is an error before assigning the returned variables. </span><span class="koboSpan" id="kobo.110.3">Therefore, the values returned with an error </span><span class="No-Break"><span class="koboSpan" id="kobo.111.1">are irrelevant.</span></span></p>
<h3><span class="koboSpan" id="kobo.112.1">Add only information that the caller does not have</span></h3>
<p><span class="koboSpan" id="kobo.113.1">When creating the error message, do not </span><a id="_idIndexMarker855"/><span class="koboSpan" id="kobo.114.1">add information that is already known by the caller of your function. </span><span class="koboSpan" id="kobo.114.2">The following example illustrates </span><span class="No-Break"><span class="koboSpan" id="kobo.115.1">this problem:</span></span></p>
<pre class="source-code"><span class="koboSpan" id="kobo.116.1">
func divide(q int, d int) (int, error) {
    if d == 0 {
        return 0, fmt.Errorf("%d can't be divided by zero", q)
    }
    return q / d, nil
}</span></pre>
<p><span class="koboSpan" id="kobo.117.1">As you can see in the preceding example, the value of </span><strong class="source-inline"><span class="koboSpan" id="kobo.118.1">q</span></strong><span class="koboSpan" id="kobo.119.1"> was returned in the error message. </span><span class="koboSpan" id="kobo.119.2">But that is not necessary, because the caller of the </span><strong class="source-inline"><span class="koboSpan" id="kobo.120.1">divide</span></strong><span class="koboSpan" id="kobo.121.1"> function has </span><span class="No-Break"><span class="koboSpan" id="kobo.122.1">this value.</span></span></p>
<p><span class="koboSpan" id="kobo.123.1">When creating errors to return in your function, do not include any arguments passed in the function, as this is known</span><a id="_idIndexMarker856"/><span class="koboSpan" id="kobo.124.1"> by the caller. </span><span class="koboSpan" id="kobo.124.2">This can lead to </span><span class="No-Break"><span class="koboSpan" id="kobo.125.1">information duplication.</span></span></p>
<h3><span class="koboSpan" id="kobo.126.1">Use lowercase and do not end with punctuation</span></h3>
<p><span class="koboSpan" id="kobo.127.1">Always use lowercase as your </span><a id="_idIndexMarker857"/><span class="koboSpan" id="kobo.128.1">error message will be concatenated with other messages when returning. </span><span class="koboSpan" id="kobo.128.2">Most of the time, you should also not use any punctuation, because the error messages will probably be linked together, and punctuation</span><a id="_idIndexMarker858"/><span class="koboSpan" id="kobo.129.1"> will end up looking odd in the middle of the </span><span class="No-Break"><span class="koboSpan" id="kobo.130.1">error message.</span></span></p>
<p><span class="koboSpan" id="kobo.131.1">The exception of the lowercase rule is when you are referring to the names of functions or methods that already have </span><span class="No-Break"><span class="koboSpan" id="kobo.132.1">capital letters.</span></span></p>
<h3><span class="koboSpan" id="kobo.133.1">Add a colon to the error message</span></h3>
<p><span class="koboSpan" id="kobo.134.1">A colon (</span><strong class="source-inline"><span class="koboSpan" id="kobo.135.1">:</span></strong><span class="koboSpan" id="kobo.136.1">) is used whenever you want to</span><a id="_idIndexMarker859"/><span class="koboSpan" id="kobo.137.1"> add information from another error message of a call made inside your code. </span><span class="koboSpan" id="kobo.137.2">Let’s use the following code as </span><span class="No-Break"><span class="koboSpan" id="kobo.138.1">an example:</span></span></p>
<pre class="source-code"><span class="koboSpan" id="kobo.139.1">
func connect(host string, conf ssh.ClientConfig) error {
    conn, err := ssh.Dial("tcp", host+":22", conf)
    if err != nil {
        return fmt.Errorf("ssh.Dial: %v", err)
    }
    . </span><span class="koboSpan" id="kobo.139.2">. </span><span class="koboSpan" id="kobo.139.3">.</span></pre>
<p><span class="koboSpan" id="kobo.140.1">In the preceding example, the </span><strong class="source-inline"><span class="koboSpan" id="kobo.141.1">connect</span></strong><span class="koboSpan" id="kobo.142.1"> function encapsulates a call to </span><strong class="source-inline"><span class="koboSpan" id="kobo.143.1">ssh.Dial</span></strong><span class="koboSpan" id="kobo.144.1">. </span><span class="koboSpan" id="kobo.144.2">We can add the context of the error to the error message by indicating which call generated the error by adding the name of the call or some information about </span><strong class="source-inline"><span class="koboSpan" id="kobo.145.1">ssh.Dial</span></strong><span class="koboSpan" id="kobo.146.1">, using a colon to separate it if necessary. </span><span class="koboSpan" id="kobo.146.2">Note that the </span><strong class="source-inline"><span class="koboSpan" id="kobo.147.1">config</span></strong><span class="koboSpan" id="kobo.148.1"> and </span><strong class="source-inline"><span class="koboSpan" id="kobo.149.1">host</span></strong><span class="koboSpan" id="kobo.150.1"> arguments are known by the caller of the </span><strong class="source-inline"><span class="koboSpan" id="kobo.151.1">connect</span></strong><span class="koboSpan" id="kobo.152.1"> function and therefore should not be added to the </span><span class="No-Break"><span class="koboSpan" id="kobo.153.1">error message.</span></span></p>
<h3><span class="koboSpan" id="kobo.154.1">Use defer, panic, and recover</span></h3>
<p><span class="koboSpan" id="kobo.155.1">Go has important mechanisms to control how the program flows during errors. </span><span class="koboSpan" id="kobo.155.2">That is mostly useful when using goroutines, because one might cause the program to crash, and you might need to be extra careful to avoid unclosed software pipes and software caches; and avoid unfreed memory and unclosed </span><span class="No-Break"><span class="koboSpan" id="kobo.156.1">file descriptors</span></span></p>
<h4><span class="koboSpan" id="kobo.157.1">defer</span></h4>
<p><span class="koboSpan" id="kobo.158.1">Go </span><strong class="source-inline"><span class="koboSpan" id="kobo.159.1">defer</span></strong><span class="koboSpan" id="kobo.160.1"> is used to push the </span><a id="_idIndexMarker860"/><span class="koboSpan" id="kobo.161.1">execution to a list that is only executed after the surrounding function has returned or when it crashes. </span><span class="koboSpan" id="kobo.161.2">The main intention of </span><strong class="source-inline"><span class="koboSpan" id="kobo.162.1">defer</span></strong><span class="koboSpan" id="kobo.163.1"> is to perform </span><a id="_idIndexMarker861"/><span class="koboSpan" id="kobo.164.1">a cleanup. </span><span class="koboSpan" id="kobo.164.2">Consider the following example of a function that copies data from one file to another and then </span><span class="No-Break"><span class="koboSpan" id="kobo.165.1">removes it:</span></span></p>
<pre class="source-code"><span class="koboSpan" id="kobo.166.1">
func moveFile(srcFile, dstFile string) error {
    src, err := os.Open(srcFile)
    if err != nil {
        return fmt.Errorf("os.Open: %v", err)
    }
    dst, err := os.Create(dstFile)
    if err != nil {
        return fmt.Errorf("os.Create: %v", err)
    }
    _, err = io.Copy(dst, src)
    if err != nil {
        return fmt.Errorf("io.Copy: %v", err)
    }
    dst.Close()
    src.Close()
    err = os.Remove(srcFile)
    if err != nil {
        return fmt.Errorf("os.Remove: %v", err)
    }
    return nil
}</span></pre>
<p><span class="koboSpan" id="kobo.167.1">In the preceding example, if an error occurs in </span><strong class="source-inline"><span class="koboSpan" id="kobo.168.1">os.Create</span></strong><span class="koboSpan" id="kobo.169.1">, the function would return before </span><a id="_idIndexMarker862"/><span class="koboSpan" id="kobo.170.1">calling </span><strong class="source-inline"><span class="koboSpan" id="kobo.171.1">src.Close()</span></strong><span class="koboSpan" id="kobo.172.1">, which means the file has not been </span><span class="No-Break"><span class="koboSpan" id="kobo.173.1">closed properly.</span></span></p>
<p><span class="koboSpan" id="kobo.174.1">The way to avoid having to </span><a id="_idIndexMarker863"/><span class="koboSpan" id="kobo.175.1">add the </span><strong class="source-inline"><span class="koboSpan" id="kobo.176.1">close</span></strong><span class="koboSpan" id="kobo.177.1"> statement repetitively throughout the code is to use </span><strong class="source-inline"><span class="koboSpan" id="kobo.178.1">defer</span></strong><span class="koboSpan" id="kobo.179.1">, like the </span><span class="No-Break"><span class="koboSpan" id="kobo.180.1">following example:</span></span></p>
<pre class="source-code"><span class="koboSpan" id="kobo.181.1">
func moveFile(srcFile, dstFile string) error {
    src, err := os.Open(srcFile)
    if err != nil {
        return fmt.Errorf("os.Open: %v", err)
    }
    defer src.Close()
    dst, err := os.Create(dstFile)
    if err != nil {
        return fmt.Errorf("os.Create: %v", err)
    }
    defer dst.Close()
    _, err = io.Copy(dst, src)
    if err != nil {
        return fmt.Errorf("io.Copy: %v", err)
    }
    err = os.Remove(srcFile)
    if err != nil {
        return fmt.Errorf("os.Remove: %v", err)
    }
    return nil
}</span></pre>
<p><span class="koboSpan" id="kobo.182.1">As you can see in the</span><a id="_idIndexMarker864"/><span class="koboSpan" id="kobo.183.1"> preceding example, </span><strong class="source-inline"><span class="koboSpan" id="kobo.184.1">defer</span></strong><span class="koboSpan" id="kobo.185.1"> is used just after a successful </span><strong class="source-inline"><span class="koboSpan" id="kobo.186.1">os.Open</span></strong><span class="koboSpan" id="kobo.187.1"> and after a successful </span><strong class="source-inline"><span class="koboSpan" id="kobo.188.1">os.Create</span></strong><span class="koboSpan" id="kobo.189.1">. </span><span class="koboSpan" id="kobo.189.2">Therefore, if there is an error, or if the function ends, it will invoke </span><strong class="source-inline"><span class="koboSpan" id="kobo.190.1">dst.Close()</span></strong><span class="koboSpan" id="kobo.191.1"> first and then </span><strong class="source-inline"><span class="koboSpan" id="kobo.192.1">src.Close()</span></strong><span class="koboSpan" id="kobo.193.1"> in a</span><a id="_idIndexMarker865"/><span class="koboSpan" id="kobo.194.1"> reverse order, like in a </span><strong class="bold"><span class="koboSpan" id="kobo.195.1">Last In, First Out</span></strong><span class="koboSpan" id="kobo.196.1"> (</span><span class="No-Break"><strong class="bold"><span class="koboSpan" id="kobo.197.1">LIFO</span></strong></span><span class="No-Break"><span class="koboSpan" id="kobo.198.1">) queue.</span></span></p>
<p><span class="koboSpan" id="kobo.199.1">Let’s see how to use </span><span class="No-Break"><strong class="source-inline"><span class="koboSpan" id="kobo.200.1">panic</span></strong></span><span class="No-Break"><span class="koboSpan" id="kobo.201.1"> now.</span></span></p>
<h4><span class="koboSpan" id="kobo.202.1">panic</span></h4>
<p><span class="koboSpan" id="kobo.203.1">When writing code, if you don’t </span><a id="_idIndexMarker866"/><span class="koboSpan" id="kobo.204.1">want to handle an error, you can use </span><strong class="source-inline"><span class="koboSpan" id="kobo.205.1">panic</span></strong><span class="koboSpan" id="kobo.206.1"> to indicate that you want to stop immediately. </span><span class="koboSpan" id="kobo.206.2">In Go, </span><strong class="source-inline"><span class="koboSpan" id="kobo.207.1">panic</span></strong><span class="koboSpan" id="kobo.208.1"> can be called in your </span><a id="_idIndexMarker867"/><span class="koboSpan" id="kobo.209.1">program by explicitly writing it, but it is also called automatically during runtime if an error occurs. </span><span class="koboSpan" id="kobo.209.2">The following is a list of the major runtime errors that </span><span class="No-Break"><span class="koboSpan" id="kobo.210.1">can occur:</span></span></p>
<ul>
<li><span class="koboSpan" id="kobo.211.1">Out-of-bounds memory access, </span><span class="No-Break"><span class="koboSpan" id="kobo.212.1">including arrays</span></span></li>
<li><span class="koboSpan" id="kobo.213.1">Assertion of a </span><span class="No-Break"><span class="koboSpan" id="kobo.214.1">wrong type</span></span></li>
<li><span class="koboSpan" id="kobo.215.1">Attempting to call a function using a variable with a </span><span class="No-Break"><strong class="source-inline"><span class="koboSpan" id="kobo.216.1">nil</span></strong></span><span class="No-Break"><span class="koboSpan" id="kobo.217.1"> pointer</span></span></li>
<li><span class="koboSpan" id="kobo.218.1">Sending data to a closed channel or </span><span class="No-Break"><span class="koboSpan" id="kobo.219.1">file descriptor</span></span></li>
<li><span class="koboSpan" id="kobo.220.1">Division </span><span class="No-Break"><span class="koboSpan" id="kobo.221.1">by zero</span></span></li>
</ul>
<p><span class="koboSpan" id="kobo.222.1">Therefore, </span><strong class="source-inline"><span class="koboSpan" id="kobo.223.1">panic</span></strong><span class="koboSpan" id="kobo.224.1"> is only used in your code when you are not planning to handle the error or are dealing with errors that are not </span><span class="No-Break"><span class="koboSpan" id="kobo.225.1">yet understood.</span></span></p>
<p><span class="koboSpan" id="kobo.226.1">It is important to note that before</span><a id="_idIndexMarker868"/><span class="koboSpan" id="kobo.227.1"> exiting the function and passing the </span><strong class="source-inline"><span class="koboSpan" id="kobo.228.1">panic</span></strong><span class="koboSpan" id="kobo.229.1"> message, the program will still run all </span><strong class="source-inline"><span class="koboSpan" id="kobo.230.1">defer</span></strong><span class="koboSpan" id="kobo.231.1"> statements that were stacked earlier in </span><span class="No-Break"><span class="koboSpan" id="kobo.232.1">the function.</span></span></p>
<p><span class="koboSpan" id="kobo.233.1">Here is an example of</span><a id="_idIndexMarker869"/><span class="koboSpan" id="kobo.234.1"> using </span><strong class="source-inline"><span class="koboSpan" id="kobo.235.1">panic</span></strong><span class="koboSpan" id="kobo.236.1"> to exit the program after receiving a negative value as </span><span class="No-Break"><span class="koboSpan" id="kobo.237.1">an argument:</span></span></p>
<pre class="source-code"><span class="koboSpan" id="kobo.238.1">
import (
    "fmt"
    "math"
)
func squareRoot(value float64) float64 {
    if value &lt; 0 {
        panic("negative values are not allowed")
    }
    return math.Sqrt(value)
}
func main() {
    fmt.Println(squareRoot(-2))
    fmt.Println("done")
}</span></pre>
<p><span class="koboSpan" id="kobo.239.1">Let’s run this program and check </span><span class="No-Break"><span class="koboSpan" id="kobo.240.1">the output:</span></span></p>
<pre class="console"><span class="koboSpan" id="kobo.241.1">
$ go run panic-example.go
panic: negative values are not allowed
goroutine 1 [running]:
main.squareRoot(...)
    Dev/Chapter07/Go/panic-example.go:10
main.main()
    Dev/Chapter07/Go/panic-example.go:17 +0x45
exit status 2</span></pre>
<p><span class="koboSpan" id="kobo.242.1">Note that the output does</span><a id="_idIndexMarker870"/><span class="koboSpan" id="kobo.243.1"> not print </span><strong class="source-inline"><span class="koboSpan" id="kobo.244.1">done</span></strong><span class="koboSpan" id="kobo.245.1">, because </span><strong class="source-inline"><span class="koboSpan" id="kobo.246.1">panic</span></strong><span class="koboSpan" id="kobo.247.1"> is called inside the </span><strong class="source-inline"><span class="koboSpan" id="kobo.248.1">squareRoot</span></strong><span class="koboSpan" id="kobo.249.1"> function, before the instruction </span><span class="No-Break"><span class="koboSpan" id="kobo.250.1">to print.</span></span></p>
<p><span class="koboSpan" id="kobo.251.1">Say we add </span><strong class="source-inline"><span class="koboSpan" id="kobo.252.1">defer</span></strong><span class="koboSpan" id="kobo.253.1"> to the function </span><a id="_idIndexMarker871"/><span class="No-Break"><span class="koboSpan" id="kobo.254.1">as follows:</span></span></p>
<pre class="source-code"><span class="koboSpan" id="kobo.255.1">
func squareRoot(value float64) float64 {
    defer fmt.Println("ending the function")
    if value &lt; 0 {
        panic("negative values are not allowed")
    }
    return math.Sqrt(value)
}</span></pre>
<p><span class="koboSpan" id="kobo.256.1">The output will be </span><span class="No-Break"><span class="koboSpan" id="kobo.257.1">as follows:</span></span></p>
<pre class="console"><span class="koboSpan" id="kobo.258.1">
$ go run panic-example.go
ending the function
panic: negative values are not allowed
goroutine 1 [running]:
main.squareRoot(…)
    Dev/Chapter07/Go/panic-example.go:10
main.main()
    Dev/Chapter07/Go/panic-example.go:17 +0x45
exit status 2</span></pre>
<p><span class="koboSpan" id="kobo.259.1">Note that the </span><strong class="source-inline"><span class="koboSpan" id="kobo.260.1">ending the function</span></strong><span class="koboSpan" id="kobo.261.1"> print statement was placed before sending the </span><strong class="source-inline"><span class="koboSpan" id="kobo.262.1">panic</span></strong><span class="koboSpan" id="kobo.263.1"> message. </span><span class="koboSpan" id="kobo.263.2">That is because, as we explained, the </span><strong class="source-inline"><span class="koboSpan" id="kobo.264.1">defer</span></strong><span class="koboSpan" id="kobo.265.1"> stack is executed before returning the </span><a id="_idIndexMarker872"/><span class="koboSpan" id="kobo.266.1">function </span><span class="No-Break"><span class="koboSpan" id="kobo.267.1">by </span></span><span class="No-Break"><strong class="source-inline"><span class="koboSpan" id="kobo.268.1">panic</span></strong></span><span class="No-Break"><span class="koboSpan" id="kobo.269.1">.</span></span></p>
<p><span class="koboSpan" id="kobo.270.1">Let’s now see how </span><a id="_idIndexMarker873"/><span class="koboSpan" id="kobo.271.1">we can </span><span class="No-Break"><span class="koboSpan" id="kobo.272.1">use </span></span><span class="No-Break"><strong class="source-inline"><span class="koboSpan" id="kobo.273.1">recover</span></strong></span><span class="No-Break"><span class="koboSpan" id="kobo.274.1">.</span></span></p>
<h4><span class="koboSpan" id="kobo.275.1">recover</span></h4>
<p><span class="koboSpan" id="kobo.276.1">In Go, </span><strong class="source-inline"><span class="koboSpan" id="kobo.277.1">recover</span></strong><span class="koboSpan" id="kobo.278.1"> is the last piece of</span><a id="_idIndexMarker874"/><span class="koboSpan" id="kobo.279.1"> error flow control necessary to</span><a id="_idIndexMarker875"/><span class="koboSpan" id="kobo.280.1"> handle errors. </span><span class="koboSpan" id="kobo.280.2">It is used to handle a </span><strong class="source-inline"><span class="koboSpan" id="kobo.281.1">panic</span></strong><span class="koboSpan" id="kobo.282.1"> situation and regain control. </span><span class="koboSpan" id="kobo.282.2">It should only be used inside a </span><strong class="source-inline"><span class="koboSpan" id="kobo.283.1">defer</span></strong><span class="koboSpan" id="kobo.284.1"> function call. </span><span class="koboSpan" id="kobo.284.2">In a normal call, </span><strong class="source-inline"><span class="koboSpan" id="kobo.285.1">recover</span></strong><span class="koboSpan" id="kobo.286.1"> will return a </span><strong class="source-inline"><span class="koboSpan" id="kobo.287.1">nil</span></strong><span class="koboSpan" id="kobo.288.1"> value, but in the </span><strong class="source-inline"><span class="koboSpan" id="kobo.289.1">panic</span></strong><span class="koboSpan" id="kobo.290.1"> situation, it will return the value given </span><span class="No-Break"><span class="koboSpan" id="kobo.291.1">to </span></span><span class="No-Break"><strong class="source-inline"><span class="koboSpan" id="kobo.292.1">panic</span></strong></span><span class="No-Break"><span class="koboSpan" id="kobo.293.1">.</span></span></p>
<p><span class="koboSpan" id="kobo.294.1">As an example, let’s consider the </span><span class="No-Break"><span class="koboSpan" id="kobo.295.1">following program:</span></span></p>
<pre class="source-code"><span class="koboSpan" id="kobo.296.1">
import "fmt"
func divide(q, d int) int {
    fmt.Println("Dividing it now")
    return q / d
}
func main() {
    fmt.Println("the division is:", divide(4, 0))
}</span></pre>
<p><span class="koboSpan" id="kobo.297.1">If you run the preceding </span><a id="_idIndexMarker876"/><span class="koboSpan" id="kobo.298.1">program, you will get the following </span><span class="No-Break"><strong class="source-inline"><span class="koboSpan" id="kobo.299.1">panic</span></strong></span><span class="No-Break"><span class="koboSpan" id="kobo.300.1"> message:</span></span></p>
<pre class="console"><span class="koboSpan" id="kobo.301.1">
$ go run division-by-zero-panic.go
Dividing it now
panic: runtime error: integer divide by zero
goroutine 1 [running]:
main.divide(...)
    Dev/Chapter07/Go/division-by-zero-panic.go:7
main.main()
    Dev/Chapter07/Go/division-by-zero-panic.go:11 +0x85
exit status 2</span></pre>
<p><span class="koboSpan" id="kobo.302.1">As you can see from this output, you have no control of the </span><strong class="source-inline"><span class="koboSpan" id="kobo.303.1">panic</span></strong><span class="koboSpan" id="kobo.304.1"> situation. </span><span class="koboSpan" id="kobo.304.2">It will basically crash the program without the possibility to handle the error properly. </span><span class="koboSpan" id="kobo.304.3">This is undesirable in most </span><a id="_idIndexMarker877"/><span class="koboSpan" id="kobo.305.1">production software, especially using </span><span class="No-Break"><span class="koboSpan" id="kobo.306.1">multiple goroutines.</span></span></p>
<p><span class="koboSpan" id="kobo.307.1">Therefore, to properly handle the </span><strong class="source-inline"><span class="koboSpan" id="kobo.308.1">panic</span></strong><span class="koboSpan" id="kobo.309.1"> situation, you should add a </span><strong class="source-inline"><span class="koboSpan" id="kobo.310.1">defer</span></strong><span class="koboSpan" id="kobo.311.1"> function to test whether it is a </span><strong class="source-inline"><span class="koboSpan" id="kobo.312.1">panic</span></strong><span class="koboSpan" id="kobo.313.1"> situation by using </span><strong class="source-inline"><span class="koboSpan" id="kobo.314.1">recover</span></strong><span class="koboSpan" id="kobo.315.1">, as in the </span><span class="No-Break"><span class="koboSpan" id="kobo.316.1">following example:</span></span></p>
<pre class="source-code"><span class="koboSpan" id="kobo.317.1">
import "fmt"
func divide(q, d int) int {
    fmt.Println("Dividing it now")
    return q / d
}
func main() {
    defer func() {
        if r := recover(); r != nil {
            fmt.Println("Got a panic:", r)
        }
    }()
    fmt.Println("the division is:", divide(4, 0))
}</span></pre>
<p><span class="koboSpan" id="kobo.318.1">After adding the </span><strong class="source-inline"><span class="koboSpan" id="kobo.319.1">defer</span></strong><span class="koboSpan" id="kobo.320.1"> function as in</span><a id="_idIndexMarker878"/><span class="koboSpan" id="kobo.321.1"> the preceding example, the output</span><a id="_idIndexMarker879"/><span class="koboSpan" id="kobo.322.1"> will be </span><span class="No-Break"><span class="koboSpan" id="kobo.323.1">as follows:</span></span></p>
<pre class="console"><span class="koboSpan" id="kobo.324.1">
$ go run division-by-zero-panic-recover.go
Dividing it now
Got a panic: runtime error: integer divide by zero</span></pre>
<p><span class="koboSpan" id="kobo.325.1">As you can see, adding a </span><strong class="source-inline"><span class="koboSpan" id="kobo.326.1">recover</span></strong><span class="koboSpan" id="kobo.327.1"> test inside the </span><strong class="source-inline"><span class="koboSpan" id="kobo.328.1">defer</span></strong><span class="koboSpan" id="kobo.329.1"> function will allow you to handle unexpected </span><strong class="source-inline"><span class="koboSpan" id="kobo.330.1">panic</span></strong><span class="koboSpan" id="kobo.331.1"> situations, avoiding your program crashing unexpectedly without doing a proper cleanup or fixing </span><span class="No-Break"><span class="koboSpan" id="kobo.332.1">the error.</span></span></p>
<p><span class="koboSpan" id="kobo.333.1">Now that we have investigated how to handle Go errors, let’s have a look at Python </span><span class="No-Break"><span class="koboSpan" id="kobo.334.1">error handling.</span></span></p>
<h2 id="_idParaDest-187"><a id="_idTextAnchor188"/><span class="koboSpan" id="kobo.335.1">Adding error handling in Python</span></h2>
<p><span class="koboSpan" id="kobo.336.1">Python handles errors differently </span><a id="_idIndexMarker880"/><span class="koboSpan" id="kobo.337.1">than Go. </span><span class="koboSpan" id="kobo.337.2">Python does not require your </span><a id="_idIndexMarker881"/><span class="koboSpan" id="kobo.338.1">function to return error values. </span><span class="koboSpan" id="kobo.338.2">In Python, errors are thrown during runtime and they are called</span><a id="_idIndexMarker882"/><span class="koboSpan" id="kobo.339.1"> exceptions. </span><span class="koboSpan" id="kobo.339.2">To handle exceptions, your code has to catch them properly and avoid </span><span class="No-Break"><span class="koboSpan" id="kobo.340.1">raising them.</span></span></p>
<h3><span class="koboSpan" id="kobo.341.1">Catching exceptions</span></h3>
<p><span class="koboSpan" id="kobo.342.1">In Python, there are built-in exceptions that are raised by many runtime errors. </span><span class="koboSpan" id="kobo.342.2">The list of built-in exceptions is quite long and can be found here: </span><a href="https://docs.python.org/3/library/exceptions.html"><span class="koboSpan" id="kobo.343.1">https://docs.python.org/3/library/exceptions.html</span></a><span class="koboSpan" id="kobo.344.1">. </span><span class="koboSpan" id="kobo.344.2">As an example, the division-by-zero error is called a </span><span class="No-Break"><strong class="source-inline"><span class="koboSpan" id="kobo.345.1">ZeroDivisionError</span></strong></span><span class="No-Break"><span class="koboSpan" id="kobo.346.1"> exception.</span></span></p>
<p><span class="koboSpan" id="kobo.347.1">To handle the error, you need to catch the </span><a id="_idIndexMarker883"/><span class="koboSpan" id="kobo.348.1">exception and then handle it by using the </span><strong class="source-inline"><span class="koboSpan" id="kobo.349.1">try</span></strong><span class="koboSpan" id="kobo.350.1">, </span><strong class="source-inline"><span class="koboSpan" id="kobo.351.1">except</span></strong><span class="koboSpan" id="kobo.352.1">, </span><strong class="source-inline"><span class="koboSpan" id="kobo.353.1">else</span></strong><span class="koboSpan" id="kobo.354.1">, and </span><strong class="source-inline"><span class="koboSpan" id="kobo.355.1">finally</span></strong><span class="koboSpan" id="kobo.356.1"> Python statements. </span><span class="koboSpan" id="kobo.356.2">To create an example of how to handle the exception for division by zero, let’s first run the following program without catching </span><span class="No-Break"><span class="koboSpan" id="kobo.357.1">an exception:</span></span></p>
<pre class="source-code"><span class="koboSpan" id="kobo.358.1">
def division(q, d):
    return q/d
print(division(1, 0))</span></pre>
<p><span class="koboSpan" id="kobo.359.1">If you run the preceding program, it will generate the </span><span class="No-Break"><span class="koboSpan" id="kobo.360.1">following output:</span></span></p>
<pre class="console"><span class="koboSpan" id="kobo.361.1">
$ python catching-division-by-zero-exception.py
Traceback (most recent call last):
  File "Chapter07/Python/catching-division-by-zero-exception.py", line 7, in &lt;module&gt;
    print(division(1, 0))
  File "Chapter07/Python/catching-division-by-zero-exception.py", line 4, in division
    return q/d
</span><strong class="bold"><span class="koboSpan" id="kobo.362.1">ZeroDivisionError</span></strong><span class="koboSpan" id="kobo.363.1">: division by zero</span></pre>
<p><span class="koboSpan" id="kobo.364.1">As you can see, the program crashes and shows the error message as </span><strong class="source-inline"><span class="koboSpan" id="kobo.365.1">Traceback</span></strong><span class="koboSpan" id="kobo.366.1"> on the screen, with details of where the error occurs and the name of the exception, in this </span><span class="No-Break"><span class="koboSpan" id="kobo.367.1">case, </span></span><span class="No-Break"><strong class="source-inline"><span class="koboSpan" id="kobo.368.1">ZeroDivisionError</span></strong></span><span class="No-Break"><span class="koboSpan" id="kobo.369.1">.</span></span></p>
<p><span class="koboSpan" id="kobo.370.1">Now, let’s update the Python code to catch this exception and handle the error more gracefully, like the </span><span class="No-Break"><span class="koboSpan" id="kobo.371.1">following code:</span></span></p>
<pre class="source-code"><span class="koboSpan" id="kobo.372.1">
def division(q, d):
    return q/d
try:
    print(division(1, 0))
except </span><strong class="bold"><span class="koboSpan" id="kobo.373.1">ZeroDivisionError</span></strong><span class="koboSpan" id="kobo.374.1">:
    print("Error: We should not divide by zero")</span></pre>
<p><span class="koboSpan" id="kobo.375.1">Now, if you run the program, it will</span><a id="_idIndexMarker884"/><span class="koboSpan" id="kobo.376.1"> print the error gracefully without crashing it, </span><span class="No-Break"><span class="koboSpan" id="kobo.377.1">as follows:</span></span></p>
<pre class="console"><span class="koboSpan" id="kobo.378.1">
$ python catching-division-by-zero-exception.py
Error: We should not divide by zero</span></pre>
<p><span class="koboSpan" id="kobo.379.1">So, whenever you think there is a possibility for the function to raise an exception by an error, use the </span><strong class="source-inline"><span class="koboSpan" id="kobo.380.1">try</span></strong><span class="koboSpan" id="kobo.381.1"> and </span><strong class="source-inline"><span class="koboSpan" id="kobo.382.1">except</span></strong><span class="koboSpan" id="kobo.383.1"> statements, as shown in the </span><span class="No-Break"><span class="koboSpan" id="kobo.384.1">preceding example.</span></span></p>
<p><span class="koboSpan" id="kobo.385.1">In addition to </span><strong class="source-inline"><span class="koboSpan" id="kobo.386.1">try</span></strong><span class="koboSpan" id="kobo.387.1"> and </span><strong class="source-inline"><span class="koboSpan" id="kobo.388.1">except</span></strong><span class="koboSpan" id="kobo.389.1"> statements, Python also allows using </span><strong class="source-inline"><span class="koboSpan" id="kobo.390.1">else</span></strong><span class="koboSpan" id="kobo.391.1"> and </span><strong class="source-inline"><span class="koboSpan" id="kobo.392.1">finally</span></strong><span class="koboSpan" id="kobo.393.1"> statements to add more flow control of the error handling. </span><span class="koboSpan" id="kobo.393.2">They are not mandatory as the flow can be controlled outside the </span><strong class="source-inline"><span class="koboSpan" id="kobo.394.1">try</span></strong><span class="koboSpan" id="kobo.395.1">/</span><strong class="source-inline"><span class="koboSpan" id="kobo.396.1">except</span></strong><span class="koboSpan" id="kobo.397.1"> statement, but they are sometimes useful. </span><span class="koboSpan" id="kobo.397.2">The following is the same example of adding the </span><strong class="source-inline"><span class="koboSpan" id="kobo.398.1">else</span></strong><span class="koboSpan" id="kobo.399.1"> and </span><span class="No-Break"><strong class="source-inline"><span class="koboSpan" id="kobo.400.1">finally</span></strong></span><span class="No-Break"><span class="koboSpan" id="kobo.401.1"> statements:</span></span></p>
<pre class="source-code"><span class="koboSpan" id="kobo.402.1">
def division(q, d):
    return q/d
try:
    result = division(10, 1)
except ZeroDivisionError:
    print("Error: We should not divide by zero")
else:
    print("Division succeded, result is:", result)
finally:
    print("done")</span></pre>
<p><span class="koboSpan" id="kobo.403.1">If you run this program, it will generate the </span><span class="No-Break"><span class="koboSpan" id="kobo.404.1">following output:</span></span></p>
<pre class="console"><span class="koboSpan" id="kobo.405.1">
$ python catch-else-finally-division-by-zero.py
Division succeded, result is: 10.0
done</span></pre>
<p><span class="koboSpan" id="kobo.406.1">Note that the </span><strong class="source-inline"><span class="koboSpan" id="kobo.407.1">else</span></strong><span class="koboSpan" id="kobo.408.1"> statement is only executed if no exception was raised in the </span><strong class="source-inline"><span class="koboSpan" id="kobo.409.1">try</span></strong><span class="koboSpan" id="kobo.410.1"> clause. </span><span class="koboSpan" id="kobo.410.2">The </span><strong class="source-inline"><span class="koboSpan" id="kobo.411.1">finally</span></strong><span class="koboSpan" id="kobo.412.1"> statement is </span><a id="_idIndexMarker885"/><span class="koboSpan" id="kobo.413.1">always executed, regardless of whether an exception was raised or not in the </span><span class="No-Break"><strong class="source-inline"><span class="koboSpan" id="kobo.414.1">try</span></strong></span><span class="No-Break"><span class="koboSpan" id="kobo.415.1"> clause.</span></span></p>
<p><span class="koboSpan" id="kobo.416.1">Now that we have seen how to catch an exception in Python, let’s discuss how to choose the exception we want </span><span class="No-Break"><span class="koboSpan" id="kobo.417.1">to catch.</span></span></p>
<h3><span class="koboSpan" id="kobo.418.1">Choosing more specific exceptions</span></h3>
<p><span class="koboSpan" id="kobo.419.1">In Python, exceptions are hierarchical </span><a id="_idIndexMarker886"/><span class="koboSpan" id="kobo.420.1">and always start with the exception called </span><strong class="source-inline"><span class="koboSpan" id="kobo.421.1">BaseException</span></strong><span class="koboSpan" id="kobo.422.1">. </span><span class="koboSpan" id="kobo.422.2">As an example, division by zero exhibits the </span><span class="No-Break"><span class="koboSpan" id="kobo.423.1">following hierarchy:</span></span></p>
<pre class="source-code"><span class="koboSpan" id="kobo.424.1">
BaseException -&gt; Exception -&gt; ArithmeticError-&gt; ZeroDivisionError</span></pre>
<p><span class="koboSpan" id="kobo.425.1">The exceptions hierarchy is quite useful, because your code can catch either a higher-level exception or a more specific one. </span><span class="koboSpan" id="kobo.425.2">With division by zero, you could catch the </span><strong class="source-inline"><span class="koboSpan" id="kobo.426.1">ArithmeticError</span></strong><span class="koboSpan" id="kobo.427.1"> exception instead of </span><strong class="source-inline"><span class="koboSpan" id="kobo.428.1">ZeroDivisionError</span></strong><span class="koboSpan" id="kobo.429.1">. </span><span class="koboSpan" id="kobo.429.2">However, it’s a good practice sometimes to catch the more specific exception instead of a </span><span class="No-Break"><span class="koboSpan" id="kobo.430.1">higher-level one.</span></span></p>
<p><span class="koboSpan" id="kobo.431.1">More specific exceptions are more desirable to be caught inside functions and libraries, because if you catch generic exceptions inside the function, you might mask the problem when another part of the code calls your function. </span><span class="koboSpan" id="kobo.431.2">So, it will depend on where you are catching and how you are </span><span class="No-Break"><span class="koboSpan" id="kobo.432.1">handling it.</span></span></p>
<p><span class="koboSpan" id="kobo.433.1">We now have a good idea how to handle errors in Go and Python. </span><span class="koboSpan" id="kobo.433.2">Let’s discuss how to add logging to </span><span class="No-Break"><span class="koboSpan" id="kobo.434.1">our code.</span></span></p>
<h1 id="_idParaDest-188"><a id="_idTextAnchor189"/><span class="koboSpan" id="kobo.435.1">Logging events</span></h1>
<p><span class="koboSpan" id="kobo.436.1">In computer software, logging is a </span><a id="_idIndexMarker887"/><span class="koboSpan" id="kobo.437.1">well-known technique used to help troubleshoot a problem, record milestones, understand behavior, retrieve information, and check historical events, among other useful actions. </span><span class="koboSpan" id="kobo.437.2">Despite these advantages, not many developers add proper logging to their code. </span><span class="koboSpan" id="kobo.437.3">In fact, some developers do nothing and add logging only when the program has problems and </span><span class="No-Break"><span class="koboSpan" id="kobo.438.1">needs debugging.</span></span></p>
<p><span class="koboSpan" id="kobo.439.1">In network automation, logging is even more important, because network elements are normally distributed and rely heavily on logging to be able to be audited in case of a problem or if an improvement is needed. </span><span class="koboSpan" id="kobo.439.2">Adding logging to your code is a good practice that will be appreciated by several levels of engineering, such as network operators, network planners, network security, and network designers, </span><span class="No-Break"><span class="koboSpan" id="kobo.440.1">among others.</span></span></p>
<p><span class="koboSpan" id="kobo.441.1">But one important point here that must be observed is that time synchronization between network elements is</span><a id="_idIndexMarker888"/><span class="koboSpan" id="kobo.442.1"> mandatory to allow logging to be useful. </span><span class="koboSpan" id="kobo.442.2">Protocols such as the </span><strong class="bold"><span class="koboSpan" id="kobo.443.1">Network Time Protocol</span></strong><span class="koboSpan" id="kobo.444.1"> (</span><strong class="bold"><span class="koboSpan" id="kobo.445.1">NTP</span></strong><span class="koboSpan" id="kobo.446.1">) or </span><strong class="bold"><span class="koboSpan" id="kobo.447.1">Precision Time Protocol</span></strong><span class="koboSpan" id="kobo.448.1"> (</span><strong class="bold"><span class="koboSpan" id="kobo.449.1">PTP</span></strong><span class="koboSpan" id="kobo.450.1">) must be used</span><a id="_idIndexMarker889"/><span class="koboSpan" id="kobo.451.1"> throughout </span><span class="No-Break"><span class="koboSpan" id="kobo.452.1">the network.</span></span></p>
<p><span class="koboSpan" id="kobo.453.1">A good practice to use logging is to use the Unix logging reference called </span><strong class="source-inline"><span class="koboSpan" id="kobo.454.1">syslog</span></strong><span class="koboSpan" id="kobo.455.1">, firstly published as an informational RFC in RFC3164 and then as a standard document in </span><span class="No-Break"><span class="koboSpan" id="kobo.456.1">RFC5424 (</span></span><a href="https://www.rfc-editor.org/rfc/rfc3164"><span class="No-Break"><span class="koboSpan" id="kobo.457.1">https://www.rfc-editor.org/rfc/rfc3164</span></span></a><span class="No-Break"><span class="koboSpan" id="kobo.458.1">).</span></span></p>
<p><span class="koboSpan" id="kobo.459.1">For our network automation code, we do not need to follow all the details in the </span><strong class="source-inline"><span class="koboSpan" id="kobo.460.1">syslog</span></strong><span class="koboSpan" id="kobo.461.1"> protocol standard, but we are going to use it as a guide for logging useful information based on </span><span class="No-Break"><span class="koboSpan" id="kobo.462.1">severity level.</span></span></p>
<p><span class="koboSpan" id="kobo.463.1">Let’s now talk about some levels of information we want when </span><span class="No-Break"><span class="koboSpan" id="kobo.464.1">logging events.</span></span></p>
<h2 id="_idParaDest-189"><a id="_idTextAnchor190"/><span class="koboSpan" id="kobo.465.1">Severity levels</span></h2>
<p><span class="koboSpan" id="kobo.466.1">The RFC5424 </span><strong class="source-inline"><span class="koboSpan" id="kobo.467.1">syslog</span></strong><span class="koboSpan" id="kobo.468.1"> protocol has defined eight levels of severity, which are described in </span><em class="italic"><span class="koboSpan" id="kobo.469.1">Section 6.2.1</span></em><span class="koboSpan" id="kobo.470.1"> of RFC5424 (</span><a href="https://www.rfc-editor.org/rfc/rfc5424#section-6.2.1"><span class="koboSpan" id="kobo.471.1">https://www.rfc-editor.org/rfc/rfc5424#section-6.2.1</span></a><span class="koboSpan" id="kobo.472.1">). </span><span class="koboSpan" id="kobo.472.2">They are mentioned in the following list </span><a id="_idIndexMarker890"/><span class="koboSpan" id="kobo.473.1">with a brief explanation of what kind of information message is intended to be added to </span><span class="No-Break"><span class="koboSpan" id="kobo.474.1">each one:</span></span></p>
<ul>
<li><strong class="source-inline"><span class="koboSpan" id="kobo.475.1">Emergency</span></strong><span class="koboSpan" id="kobo.476.1">: The system is not operational and there is no </span><span class="No-Break"><span class="koboSpan" id="kobo.477.1">possible recovery.</span></span></li>
<li><strong class="source-inline"><span class="koboSpan" id="kobo.478.1">Alert</span></strong><span class="koboSpan" id="kobo.479.1">: Immediate attention </span><span class="No-Break"><span class="koboSpan" id="kobo.480.1">is required.</span></span></li>
<li><strong class="source-inline"><span class="koboSpan" id="kobo.481.1">Critical</span></strong><span class="koboSpan" id="kobo.482.1">: Something bad is happening and quick attention is required to </span><span class="No-Break"><span class="koboSpan" id="kobo.483.1">fix it.</span></span></li>
<li><strong class="source-inline"><span class="koboSpan" id="kobo.484.1">Error</span></strong><span class="koboSpan" id="kobo.485.1">: Failure is occurring but does not need </span><span class="No-Break"><span class="koboSpan" id="kobo.486.1">urgent attention.</span></span></li>
<li><strong class="source-inline"><span class="koboSpan" id="kobo.487.1">Warning or Warn</span></strong><span class="koboSpan" id="kobo.488.1">: Indicates that something is wrong and might cause an error in the future, such as software </span><span class="No-Break"><span class="koboSpan" id="kobo.489.1">not updated.</span></span></li>
<li><strong class="source-inline"><span class="koboSpan" id="kobo.490.1">Notice</span></strong><span class="koboSpan" id="kobo.491.1">: An important milestone that has been reached and might indicate a future warning, such as</span><a id="_idIndexMarker891"/><span class="koboSpan" id="kobo.492.1"> configuration not saved or resource utilization limit </span><span class="No-Break"><span class="koboSpan" id="kobo.493.1">not set.</span></span></li>
<li><strong class="source-inline"><span class="koboSpan" id="kobo.494.1">Informational or Info</span></strong><span class="koboSpan" id="kobo.495.1">: Normal operational milestone messages. </span><span class="koboSpan" id="kobo.495.2">Used later for audit </span><span class="No-Break"><span class="koboSpan" id="kobo.496.1">and investigation.</span></span></li>
<li><strong class="source-inline"><span class="koboSpan" id="kobo.497.1">Debug</span></strong><span class="koboSpan" id="kobo.498.1">: Used by developers to troubleshoot a problem or to investigate </span><span class="No-Break"><span class="koboSpan" id="kobo.499.1">possible improvements.</span></span></li>
</ul>
<p><span class="koboSpan" id="kobo.500.1">Although these eight levels are defined in the </span><strong class="source-inline"><span class="koboSpan" id="kobo.501.1">syslog</span></strong><span class="koboSpan" id="kobo.502.1"> protocol, they are quite ambiguous and open to different interpretations. </span><span class="koboSpan" id="kobo.502.2">For instance, </span><strong class="source-inline"><span class="koboSpan" id="kobo.503.1">Alert</span></strong><span class="koboSpan" id="kobo.504.1"> and </span><strong class="source-inline"><span class="koboSpan" id="kobo.505.1">Emergency</span></strong><span class="koboSpan" id="kobo.506.1"> might be different for different developers when writing code, as is the case with other levels, such as </span><strong class="source-inline"><span class="koboSpan" id="kobo.507.1">Notice</span></strong><span class="koboSpan" id="kobo.508.1"> and </span><strong class="source-inline"><span class="koboSpan" id="kobo.509.1">Informational</span></strong><span class="koboSpan" id="kobo.510.1">. </span><span class="koboSpan" id="kobo.510.2">Therefore, some network developers prefer to use less levels with easier interpretation. </span><span class="koboSpan" id="kobo.510.3">The number will depend on how the network operates, but varies between three and five levels. </span><span class="koboSpan" id="kobo.510.4">For Go and Python, the number of levels will depend on the library you are using to create the log messages. </span><span class="koboSpan" id="kobo.510.5">Some might have more levels </span><span class="No-Break"><span class="koboSpan" id="kobo.511.1">than others.</span></span></p>
<p><span class="koboSpan" id="kobo.512.1">Now, let’s investigate how to add log events to your code using Go </span><span class="No-Break"><span class="koboSpan" id="kobo.513.1">and Python.</span></span></p>
<h1 id="_idParaDest-190"><a id="_idTextAnchor191"/><span class="koboSpan" id="kobo.514.1">Adding logging to your code</span></h1>
<p><span class="koboSpan" id="kobo.515.1">Adding event logging in your code </span><a id="_idIndexMarker892"/><span class="koboSpan" id="kobo.516.1">will be different in Go and in Python and will vary depending on the library used in your code. </span><span class="koboSpan" id="kobo.516.2">But the idea for both languages is to divide the information into severity levels, as done </span><span class="No-Break"><span class="koboSpan" id="kobo.517.1">in </span></span><span class="No-Break"><strong class="source-inline"><span class="koboSpan" id="kobo.518.1">syslog</span></strong></span><span class="No-Break"><span class="koboSpan" id="kobo.519.1">.</span></span></p>
<p><span class="koboSpan" id="kobo.520.1">The severity log levels will also vary depending on the library used. </span><span class="koboSpan" id="kobo.520.2">Python and Go have standard log libraries, but you are also able to use third-party libraries to log events in </span><span class="No-Break"><span class="koboSpan" id="kobo.521.1">both languages.</span></span></p>
<p><span class="koboSpan" id="kobo.522.1">One important point here is that when writing the code, you will decide whether there is a need to add a logging event line. </span><span class="koboSpan" id="kobo.522.2">The line of logging added must carry some information that will signal to the program which level of severity the message is. </span><span class="koboSpan" id="kobo.522.3">Therefore, important messages such as failures will have priority over less-important messages, such as debugging. </span><span class="koboSpan" id="kobo.522.4">Ideally, the decision about which level of logging should be exposed is normally made by</span><a id="_idIndexMarker893"/><span class="koboSpan" id="kobo.523.1"> adding input arguments to your program that allow setting the log level. </span><span class="koboSpan" id="kobo.523.2">So, if you are running the program for debugging, it will generate much more information compared to a </span><span class="No-Break"><span class="koboSpan" id="kobo.524.1">normal operation.</span></span></p>
<p><span class="koboSpan" id="kobo.525.1">Let’s now see how we can add logging events to our code in Go and then we check how to do it </span><span class="No-Break"><span class="koboSpan" id="kobo.526.1">in Python.</span></span></p>
<h2 id="_idParaDest-191"><a id="_idTextAnchor192"/><span class="koboSpan" id="kobo.527.1">Adding event logging in Go</span></h2>
<p><span class="koboSpan" id="kobo.528.1">The Go language has a standard </span><a id="_idIndexMarker894"/><span class="koboSpan" id="kobo.529.1">library for logging that comes with the Go installation, but it is quite</span><a id="_idIndexMarker895"/><span class="koboSpan" id="kobo.530.1"> limited. </span><span class="koboSpan" id="kobo.530.2">For more advanced logging capabilities in Go, you might want to use third-party libraries </span><span class="No-Break"><span class="koboSpan" id="kobo.531.1">for logging.</span></span></p>
<p><span class="koboSpan" id="kobo.532.1">Let’s see how we can</span><a id="_idIndexMarker896"/><span class="koboSpan" id="kobo.533.1"> use the standard library and then check other popular </span><span class="No-Break"><span class="koboSpan" id="kobo.534.1">third-party libraries.</span></span></p>
<h3><span class="koboSpan" id="kobo.535.1">Using standard Go logging</span></h3>
<p><span class="koboSpan" id="kobo.536.1">The standard logging</span><a id="_idIndexMarker897"/><span class="koboSpan" id="kobo.537.1"> library in Go can be imported using </span><strong class="source-inline"><span class="koboSpan" id="kobo.538.1">log</span></strong><span class="koboSpan" id="kobo.539.1"> in the </span><strong class="source-inline"><span class="koboSpan" id="kobo.540.1">import</span></strong><span class="koboSpan" id="kobo.541.1"> statement. </span><span class="koboSpan" id="kobo.541.2">By default, Go standard</span><a id="_idIndexMarker898"/><span class="koboSpan" id="kobo.542.1"> logging does not provide any severity levels, but it has some helper functions that can help create logs. </span><span class="koboSpan" id="kobo.542.2">The helper functions are </span><span class="No-Break"><span class="koboSpan" id="kobo.543.1">listed here:</span></span></p>
<ul>
<li><strong class="source-inline"><span class="koboSpan" id="kobo.544.1">Print</span></strong><span class="koboSpan" id="kobo.545.1">, </span><strong class="source-inline"><span class="koboSpan" id="kobo.546.1">Printf</span></strong><span class="koboSpan" id="kobo.547.1">, and </span><strong class="source-inline"><span class="koboSpan" id="kobo.548.1">Println</span></strong><span class="koboSpan" id="kobo.549.1">: These functions print the message passed to them in the terminal </span><span class="No-Break"><span class="koboSpan" id="kobo.550.1">using </span></span><span class="No-Break"><strong class="source-inline"><span class="koboSpan" id="kobo.551.1">stderr</span></strong></span></li>
<li><strong class="source-inline"><span class="koboSpan" id="kobo.552.1">Panic</span></strong><span class="koboSpan" id="kobo.553.1">, </span><strong class="source-inline"><span class="koboSpan" id="kobo.554.1">Panicf</span></strong><span class="koboSpan" id="kobo.555.1">, and </span><strong class="source-inline"><span class="koboSpan" id="kobo.556.1">Panicln</span></strong><span class="koboSpan" id="kobo.557.1">: These are like </span><strong class="source-inline"><span class="koboSpan" id="kobo.558.1">Print</span></strong><span class="koboSpan" id="kobo.559.1">, but they call </span><strong class="source-inline"><span class="koboSpan" id="kobo.560.1">Panic</span></strong><span class="koboSpan" id="kobo.561.1"> after printing the </span><span class="No-Break"><span class="koboSpan" id="kobo.562.1">log message</span></span></li>
<li><strong class="source-inline"><span class="koboSpan" id="kobo.563.1">Fatal</span></strong><span class="koboSpan" id="kobo.564.1">, </span><strong class="source-inline"><span class="koboSpan" id="kobo.565.1">Fatalf</span></strong><span class="koboSpan" id="kobo.566.1">, and </span><strong class="source-inline"><span class="koboSpan" id="kobo.567.1">Fatalln</span></strong><span class="koboSpan" id="kobo.568.1">: These also work like </span><strong class="source-inline"><span class="koboSpan" id="kobo.569.1">Print</span></strong><span class="koboSpan" id="kobo.570.1">, but they call </span><strong class="source-inline"><span class="koboSpan" id="kobo.571.1">os.Exit(1)</span></strong><span class="koboSpan" id="kobo.572.1"> after printing the </span><span class="No-Break"><span class="koboSpan" id="kobo.573.1">log message</span></span></li>
</ul>
<p><span class="koboSpan" id="kobo.574.1">The following is a simple example of using the standard Go </span><span class="No-Break"><span class="koboSpan" id="kobo.575.1">logging library:</span></span></p>
<pre class="source-code"><span class="koboSpan" id="kobo.576.1">
import (
    "log"
    "os/user"
)
func main() {
    user, err := user.Current()
    if err != nil {
        log.Fatalf("Failed with error: %v", err)
    }
    log.Printf("Current user is %s", user.Username)
}</span></pre>
<p><span class="koboSpan" id="kobo.577.1">Running this program will print</span><a id="_idIndexMarker899"/><span class="koboSpan" id="kobo.578.1"> the following output </span><span class="No-Break"><span class="koboSpan" id="kobo.579.1">without errors:</span></span></p>
<pre class="console"><span class="koboSpan" id="kobo.580.1">
% go run standard-logging.go
2022/11/08 18:53:24 Current user is claus</span></pre>
<p><span class="koboSpan" id="kobo.581.1">If for any reason it is not possible to retrieve the current user, it will call </span><strong class="source-inline"><span class="koboSpan" id="kobo.582.1">Fatalf</span></strong><span class="koboSpan" id="kobo.583.1">, which will call </span><strong class="source-inline"><span class="koboSpan" id="kobo.584.1">os.Exit(1)</span></strong><span class="koboSpan" id="kobo.585.1"> after </span><a id="_idIndexMarker900"/><span class="koboSpan" id="kobo.586.1">printing the </span><span class="No-Break"><span class="koboSpan" id="kobo.587.1">failed message.</span></span></p>
<p><span class="koboSpan" id="kobo.588.1">Now, let’s show a more complex example of how to create severity levels using the standard logging library and saving it to </span><span class="No-Break"><span class="koboSpan" id="kobo.589.1">a file:</span></span></p>
<pre class="source-code"><span class="koboSpan" id="kobo.590.1">
import (
    "log"
    "os"
)
var criticalLog, errorLog, warnLog, infoLog, debugLog *log.Logger
func init() {
    file, err := os.Create("log-file.txt")
    if err != nil {
        log.Fatal(err)
    }
    flags := log.Ldate | log.Ltime
    criticalLog = log.New(file, "CRITICAL: ", flags)
    errorLog = log.New(file, "ERROR: ", flags)
    warnLog = log.New(file, "WARNING: ", flags)
    infoLog = log.New(file, "INFO: ", flags)
    debugLog = log.New(file, "DEBUG: ", flags)
}
func main() {
    infoLog.Print("That is a milestone")
    errorLog.Print("Got an error here")
    debugLog.Print("Extra information for a debug")
    warnLog.Print("You should be warned about this")
}</span></pre>
<p><span class="koboSpan" id="kobo.591.1">In the preceding example, we</span><a id="_idIndexMarker901"/><span class="koboSpan" id="kobo.592.1"> created five levels of severity that can be used to write to a file as required. </span><span class="koboSpan" id="kobo.592.2">Note that in Go, the </span><strong class="source-inline"><span class="koboSpan" id="kobo.593.1">init()</span></strong><span class="koboSpan" id="kobo.594.1"> function is executed before the </span><strong class="source-inline"><span class="koboSpan" id="kobo.595.1">main()</span></strong><span class="koboSpan" id="kobo.596.1"> function. </span><span class="koboSpan" id="kobo.596.2">If you want to use</span><a id="_idIndexMarker902"/><span class="koboSpan" id="kobo.597.1"> these log definitions within other packages, remember to use capitalization of the variables; otherwise, the variables will be local to this package; for example, </span><strong class="source-inline"><span class="koboSpan" id="kobo.598.1">errorLog</span></strong><span class="koboSpan" id="kobo.599.1"> should </span><span class="No-Break"><span class="koboSpan" id="kobo.600.1">be </span></span><span class="No-Break"><strong class="source-inline"><span class="koboSpan" id="kobo.601.1">ErrorLog</span></strong></span><span class="No-Break"><span class="koboSpan" id="kobo.602.1">.</span></span></p>
<p><span class="koboSpan" id="kobo.603.1">Also, if you want to set a log level to avoid </span><strong class="source-inline"><span class="koboSpan" id="kobo.604.1">Debug</span></strong><span class="koboSpan" id="kobo.605.1"> or </span><strong class="source-inline"><span class="koboSpan" id="kobo.606.1">Info</span></strong><span class="koboSpan" id="kobo.607.1"> messages, you will have to pass an argument to your program and suppress lower levels of severity depending on the level set. </span><span class="koboSpan" id="kobo.607.2">Using the Go</span><a id="_idIndexMarker903"/><span class="koboSpan" id="kobo.608.1"> standard logging library, you will have to do that on </span><span class="No-Break"><span class="koboSpan" id="kobo.609.1">your own.</span></span></p>
<p><span class="koboSpan" id="kobo.610.1">Now, let’s investigate a </span><a id="_idIndexMarker904"/><span class="koboSpan" id="kobo.611.1">third-party logging library that is very popular with </span><span class="No-Break"><span class="koboSpan" id="kobo.612.1">Go developers.</span></span></p>
<h3><span class="koboSpan" id="kobo.613.1">Using logrus</span></h3>
<p><span class="koboSpan" id="kobo.614.1">Perhaps one of the most popular libraries for logging in Go, </span><strong class="source-inline"><span class="koboSpan" id="kobo.615.1">logrus</span></strong><span class="koboSpan" id="kobo.616.1"> is a structured logging library with several </span><a id="_idIndexMarker905"/><span class="koboSpan" id="kobo.617.1">logging capabilities. </span><strong class="source-inline"><span class="koboSpan" id="kobo.618.1">logrus</span></strong><span class="koboSpan" id="kobo.619.1"> has seven log levels and is compatible with the standard logging library. </span><span class="koboSpan" id="kobo.619.2">By default, the library allows you </span><a id="_idIndexMarker906"/><span class="koboSpan" id="kobo.620.1">to set the log level, so it won’t create noise if you don’t want to see </span><span class="No-Break"><span class="koboSpan" id="kobo.621.1">debugging information.</span></span></p>
<p><span class="koboSpan" id="kobo.622.1">Here is a simple example of using </span><strong class="source-inline"><span class="koboSpan" id="kobo.623.1">logrus</span></strong><span class="koboSpan" id="kobo.624.1"> and setting the log level to </span><strong class="source-inline"><span class="koboSpan" id="kobo.625.1">Error</span></strong><span class="koboSpan" id="kobo.626.1">, which means lower-level logs will not show, such as </span><strong class="source-inline"><span class="koboSpan" id="kobo.627.1">Warning</span></strong><span class="koboSpan" id="kobo.628.1">, </span><strong class="source-inline"><span class="koboSpan" id="kobo.629.1">Info,</span></strong> <span class="No-Break"><span class="koboSpan" id="kobo.630.1">or </span></span><span class="No-Break"><strong class="source-inline"><span class="koboSpan" id="kobo.631.1">Debug</span></strong></span><span class="No-Break"><span class="koboSpan" id="kobo.632.1">:</span></span></p>
<pre class="source-code"><span class="koboSpan" id="kobo.633.1">
import (
    log "github.com/sirupsen/logrus"
)
func init() {
    log.SetFormatter(&amp;log.TextFormatter{
        DisableColors: true,
        FullTimestamp: true,
    })
    </span><strong class="bold"><span class="koboSpan" id="kobo.634.1">log.SetLevel(log.ErrorLevel)</span></strong><span class="koboSpan" id="kobo.635.1">
}
func main() {
    log.Debug("Debug is suppressed in error level")
    log.Info("This info won't show in error level")
    log.Error("Got an error here")
}</span></pre>
<p><span class="koboSpan" id="kobo.636.1">Running the preceding example</span><a id="_idIndexMarker907"/><span class="koboSpan" id="kobo.637.1"> will only show the following output in </span><a id="_idIndexMarker908"/><span class="No-Break"><span class="koboSpan" id="kobo.638.1">the terminal:</span></span></p>
<pre class="console"><span class="koboSpan" id="kobo.639.1">
% go run logrus-logging.go
time="2022-11-09T11:16:48-03:00" level=error msg="Got an error here"</span></pre>
<p><span class="koboSpan" id="kobo.640.1">Because the level of severity is set to </span><strong class="source-inline"><span class="koboSpan" id="kobo.641.1">ErrorLevel</span></strong><span class="koboSpan" id="kobo.642.1">, none of the less-significant log messages will be displayed – in the example, the calls for </span><strong class="source-inline"><span class="koboSpan" id="kobo.643.1">log.Info</span></strong> <span class="No-Break"><span class="koboSpan" id="kobo.644.1">and </span></span><span class="No-Break"><strong class="source-inline"><span class="koboSpan" id="kobo.645.1">log.Debug</span></strong></span><span class="No-Break"><span class="koboSpan" id="kobo.646.1">.</span></span></p>
<p><strong class="source-inline"><span class="koboSpan" id="kobo.647.1">logrus</span></strong><span class="koboSpan" id="kobo.648.1"> is very flexible and powerful with plenty of examples of usage on the internet. </span><span class="koboSpan" id="kobo.648.2">More details on </span><strong class="source-inline"><span class="koboSpan" id="kobo.649.1">logrus</span></strong><span class="koboSpan" id="kobo.650.1"> can be found </span><span class="No-Break"><span class="koboSpan" id="kobo.651.1">at </span></span><a href="https://github.com/sirupsen/logrus"><span class="No-Break"><span class="koboSpan" id="kobo.652.1">https://github.com/sirupsen/logrus</span></span></a><span class="No-Break"><span class="koboSpan" id="kobo.653.1">.</span></span></p>
<p><span class="koboSpan" id="kobo.654.1">If you want to use more</span><a id="_idIndexMarker909"/><span class="koboSpan" id="kobo.655.1"> logging libraries in Go, here is a compiled list of third-party log </span><span class="No-Break"><span class="koboSpan" id="kobo.656.1">libraries: </span></span><a href="https://awesome-go.com/logging/"><span class="No-Break"><span class="koboSpan" id="kobo.657.1">https://awesome-go.com/logging/</span></span></a><span class="No-Break"><span class="koboSpan" id="kobo.658.1">.</span></span></p>
<p><span class="koboSpan" id="kobo.659.1">Now, let’s check how we add logging to our code </span><span class="No-Break"><span class="koboSpan" id="kobo.660.1">using Python.</span></span></p>
<h2 id="_idParaDest-192"><a id="_idTextAnchor193"/><span class="koboSpan" id="kobo.661.1">Add event logging in Python</span></h2>
<p><span class="koboSpan" id="kobo.662.1">In comparison to Go, Python</span><a id="_idIndexMarker910"/><span class="koboSpan" id="kobo.663.1"> has added many more features to the standard logging library. </span><span class="koboSpan" id="kobo.663.2">Despite the better support, the Python community has also developed </span><a id="_idIndexMarker911"/><span class="koboSpan" id="kobo.664.1">many third-party </span><span class="No-Break"><span class="koboSpan" id="kobo.665.1">logging libraries.</span></span></p>
<p><span class="koboSpan" id="kobo.666.1">Let’s have a look at the standard library and then a popular third-party library </span><span class="No-Break"><span class="koboSpan" id="kobo.667.1">for Python.</span></span></p>
<h3><span class="koboSpan" id="kobo.668.1">Using standard logging for Python</span></h3>
<p><span class="koboSpan" id="kobo.669.1">The standard logging library comes with five severity levels and an extra level that is used to indicate the level was</span><a id="_idIndexMarker912"/><span class="koboSpan" id="kobo.670.1"> not set by the logger. </span><span class="koboSpan" id="kobo.670.2">Each level has a number associated with it that can be </span><a id="_idIndexMarker913"/><span class="koboSpan" id="kobo.671.1">used to interpret the priority level, where lower numbers have lower priority. </span><span class="koboSpan" id="kobo.671.2">The levels are </span><strong class="source-inline"><span class="koboSpan" id="kobo.672.1">CRITICAL</span></strong><span class="koboSpan" id="kobo.673.1"> (50), </span><strong class="source-inline"><span class="koboSpan" id="kobo.674.1">ERROR</span></strong><span class="koboSpan" id="kobo.675.1"> (40), </span><strong class="source-inline"><span class="koboSpan" id="kobo.676.1">WARNING</span></strong><span class="koboSpan" id="kobo.677.1"> (30), </span><strong class="source-inline"><span class="koboSpan" id="kobo.678.1">INFO</span></strong><span class="koboSpan" id="kobo.679.1"> (20), </span><strong class="source-inline"><span class="koboSpan" id="kobo.680.1">DEBUG</span></strong><span class="koboSpan" id="kobo.681.1"> (10), and </span><span class="No-Break"><strong class="source-inline"><span class="koboSpan" id="kobo.682.1">NOTSET</span></strong></span><span class="No-Break"><span class="koboSpan" id="kobo.683.1"> (0).</span></span></p>
<p><span class="koboSpan" id="kobo.684.1">The </span><strong class="source-inline"><span class="koboSpan" id="kobo.685.1">NOTSET</span></strong><span class="koboSpan" id="kobo.686.1"> level is useful when using log hierarchies, allowing a non-root logger to delegate the level to </span><span class="No-Break"><span class="koboSpan" id="kobo.687.1">its parent.</span></span></p>
<p><span class="koboSpan" id="kobo.688.1">The following is an example of using Python </span><span class="No-Break"><span class="koboSpan" id="kobo.689.1">standard logging:</span></span></p>
<pre class="source-code"><span class="koboSpan" id="kobo.690.1">
import logging
    logging.basicConfig(
        filename='file-log.txt',
        level=</span><strong class="bold"><span class="koboSpan" id="kobo.691.1">logging.ERROR</span></strong><span class="koboSpan" id="kobo.692.1">,
        format='%(asctime)s.%(msecs)03d %(levelname)s: %(message)s',
        datefmt='%Y-%m-%d %H:%M:%S',
)
logging.debug("This won't show, level is set to info")
logging.info("Info is not that important as well")
logging.warning("Warning will not show as well")
logging.error("This is an error")</span></pre>
<p><span class="koboSpan" id="kobo.693.1">Running the preceding program will produce the following line in the output file </span><span class="No-Break"><span class="koboSpan" id="kobo.694.1">called </span></span><span class="No-Break"><strong class="source-inline"><span class="koboSpan" id="kobo.695.1">file-log.txt</span></strong></span><span class="No-Break"><span class="koboSpan" id="kobo.696.1">:</span></span></p>
<pre class="console"><span class="koboSpan" id="kobo.697.1">
2022-11-09 14:48:52.920 ERROR: This is an error</span></pre>
<p><span class="koboSpan" id="kobo.698.1">As illustrated in the preceding code, setting the level to </span><strong class="source-inline"><span class="koboSpan" id="kobo.699.1">logging.ERROR</span></strong><span class="koboSpan" id="kobo.700.1"> will not allow the lower-level log messages to be written in the file. </span><span class="koboSpan" id="kobo.700.2">The program just ignores the </span><strong class="source-inline"><span class="koboSpan" id="kobo.701.1">logging.debug()</span></strong><span class="koboSpan" id="kobo.702.1">, </span><strong class="source-inline"><span class="koboSpan" id="kobo.703.1">logging.info()</span></strong><span class="koboSpan" id="kobo.704.1">, and </span><span class="No-Break"><strong class="source-inline"><span class="koboSpan" id="kobo.705.1">logging.warning()</span></strong></span><span class="No-Break"><span class="koboSpan" id="kobo.706.1"> calls.</span></span></p>
<p><span class="koboSpan" id="kobo.707.1">Another important point is to show how easier it is to use standard logging in Python. </span><span class="koboSpan" id="kobo.707.2">The preceding example shows that you just need one call to </span><strong class="source-inline"><span class="koboSpan" id="kobo.708.1">logging.basicConfig</span></strong><span class="koboSpan" id="kobo.709.1"> to set almost everything you need, from the formatter to the level </span><span class="No-Break"><span class="koboSpan" id="kobo.710.1">of severity.</span></span></p>
<p><span class="koboSpan" id="kobo.711.1">In addition to being easy to use, the Python community has created great tutorials and documentation for the standard</span><a id="_idIndexMarker914"/><span class="koboSpan" id="kobo.712.1"> logging library. </span><span class="koboSpan" id="kobo.712.2">Here are the three main references for the documentation and advanced </span><span class="No-Break"><span class="koboSpan" id="kobo.713.1">utilization info:</span></span></p>
<ul>
<li><a href="https://docs.python.org/3/library/logging"><span class="No-Break"><span class="koboSpan" id="kobo.714.1">https://docs.python.org/3/library/logging</span></span></a></li>
<li><a href="https://docs.python.org/3/howto/logging.html"><span class="No-Break"><span class="koboSpan" id="kobo.715.1">https://docs.python.org/3/howto/logging.html</span></span></a></li>
<li><a href="https://docs.python.org/3/howto/logging-cookbook.html"><span class="No-Break"><span class="koboSpan" id="kobo.716.1">https://docs.python.org/3/howto/logging-cookbook.html</span></span></a></li>
</ul>
<p><span class="koboSpan" id="kobo.717.1">In essence, the Python standard logging library is quite complete, and you won’t need to use a third-party library for </span><a id="_idIndexMarker915"/><span class="koboSpan" id="kobo.718.1">most of your work. </span><span class="koboSpan" id="kobo.718.2">However, there are some nice and interesting features in one popular third-party library for Python called </span><strong class="source-inline"><span class="koboSpan" id="kobo.719.1">loguru</span></strong><span class="koboSpan" id="kobo.720.1">. </span><span class="koboSpan" id="kobo.720.2">Let’s see how to </span><span class="No-Break"><span class="koboSpan" id="kobo.721.1">use it.</span></span></p>
<h3><span class="koboSpan" id="kobo.722.1">Using Python loguru</span></h3>
<p><span class="koboSpan" id="kobo.723.1">Python </span><strong class="source-inline"><span class="koboSpan" id="kobo.724.1">loguru</span></strong><span class="koboSpan" id="kobo.725.1"> provides a few more features than the standard Python logging library and has the aim of being easier to use</span><a id="_idIndexMarker916"/><span class="koboSpan" id="kobo.726.1"> and configure. </span><span class="koboSpan" id="kobo.726.2">For instance, using </span><strong class="source-inline"><span class="koboSpan" id="kobo.727.1">loguru</span></strong><span class="koboSpan" id="kobo.728.1">, you will be able to set file rotation on the log file, use a more advanced string formatter, and use decorators to catch exceptions on functions, and it is thread and </span><a id="_idIndexMarker917"/><span class="No-Break"><span class="koboSpan" id="kobo.729.1">multiprocessing safe.</span></span></p>
<p><span class="koboSpan" id="kobo.730.1">It also has interesting features that can allow you to add extra information to the logging by using a </span><strong class="source-inline"><span class="koboSpan" id="kobo.731.1">patch</span></strong><span class="koboSpan" id="kobo.732.1"> method (more on the </span><strong class="source-inline"><span class="koboSpan" id="kobo.733.1">patch</span></strong><span class="koboSpan" id="kobo.734.1"> method </span><span class="No-Break"><span class="koboSpan" id="kobo.735.1">at </span></span><a href="https://loguru.readthedocs.io/en/stable/api/logger.html#loguru._logger.Logger.patch"><span class="No-Break"><span class="koboSpan" id="kobo.736.1">https://loguru.readthedocs.io/en/stable/api/logger.html#loguru._logger.Logger.patch</span></span></a><span class="No-Break"><span class="koboSpan" id="kobo.737.1">).</span></span></p>
<p><span class="koboSpan" id="kobo.738.1">The following is a simple example </span><span class="No-Break"><span class="koboSpan" id="kobo.739.1">using </span></span><span class="No-Break"><strong class="source-inline"><span class="koboSpan" id="kobo.740.1">loguru</span></strong></span><span class="No-Break"><span class="koboSpan" id="kobo.741.1">:</span></span></p>
<pre class="source-code"><span class="koboSpan" id="kobo.742.1">
from loguru import logger
logger.add(
    "file-log-{time}.txt",
    rotation="1 MB",
    colorize=False,
    level="ERROR",
)
logger.debug("That's not going to show")
logger.warning("This will not show")
logger.error("Got an error")</span></pre>
<p><span class="koboSpan" id="kobo.743.1">Running the preceding example </span><a id="_idIndexMarker918"/><span class="koboSpan" id="kobo.744.1">will create a file with the date and time with log messages that will rotate if it reaches 1 MB in size. </span><span class="koboSpan" id="kobo.744.2">The output</span><a id="_idIndexMarker919"/><span class="koboSpan" id="kobo.745.1"> written in the file will look </span><span class="No-Break"><span class="koboSpan" id="kobo.746.1">as follows:</span></span></p>
<pre class="console"><span class="koboSpan" id="kobo.747.1">
% cat file-log-2022-11-09_15-53-58_056790.txt
2022-11-09 15:53:58.063 | ERROR    | __main__:&lt;module&gt;:13 - Got an error</span></pre>
<p><span class="koboSpan" id="kobo.748.1">More detailed documentation can be found at </span><a href="https://loguru.readthedocs.io"><span class="koboSpan" id="kobo.749.1">https://loguru.readthedocs.io</span></a><span class="koboSpan" id="kobo.750.1"> and the source code </span><span class="No-Break"><span class="koboSpan" id="kobo.751.1">at </span></span><a href="https://github.com/Delgan/loguru"><span class="No-Break"><span class="koboSpan" id="kobo.752.1">https://github.com/Delgan/loguru</span></span></a><span class="No-Break"><span class="koboSpan" id="kobo.753.1">.</span></span></p>
<h1 id="_idParaDest-193"><a id="_idTextAnchor194"/><span class="koboSpan" id="kobo.754.1">Summary</span></h1>
<p><span class="koboSpan" id="kobo.755.1">After reading this chapter, you are probably more aware of why we need to handle errors and why we need to create proper event logging. </span><span class="koboSpan" id="kobo.755.2">You also should be more familiar with the differences between how Go and Python handle errors. </span><span class="koboSpan" id="kobo.755.3">Furthermore, you saw the differences in how to use standard libraries and third-party libraries for event logging. </span><span class="koboSpan" id="kobo.755.4">From now on, your network automation code design will have a special section on logging and </span><span class="No-Break"><span class="koboSpan" id="kobo.756.1">error handling.</span></span></p>
<p><span class="koboSpan" id="kobo.757.1">In the next chapter, we are going to talk about how we can scale our code and how our network automation solution can interact with </span><span class="No-Break"><span class="koboSpan" id="kobo.758.1">large networks.</span></span></p>
</div>
</body></html>